mllib/src/main/scala/org/apache/spark/ml/tree/impl/LocalDecisionTreeUtils.scala

```diff
@@ -0,0 +1,259 @@
+   * columns.  The number of partitions will be min(input RDD's number of partitions, numColumns).
+   *
+   * @param rowStore  An array of input data rows, each represented as an
+   *                  int array of binned feature values
+   * @return Transpose of rowStore with
+   *
+   * TODO: Add implementation for sparse data.
+   *       For sparse data, distribute more evenly based on number of non-zeros.
+   *       (First collect stats to decide how to partition.)
+   */
+  private[impl] def rowToColumnStoreDense(rowStore: Array[Array[Int]]): Array[Array[Int]] = {
+    // Compute the number of rows in the data
+    val numRows = {
+      val longNumRows: Long = rowStore.size
+      require(longNumRows < Int.MaxValue, s"rowToColumnStore given RDD with $longNumRows rows," +
+        s" but can handle at most ${Int.MaxValue} rows")
+      longNumRows.toInt
+    }
+
+    // Return an empty array for a dataset with zero rows or columns, otherwise
+    // return the transpose of the rowStore matrix
+    if (numRows == 0 || rowStore(0).size == 0) {
+      Array.empty
+    } else {
+      val numCols = rowStore(0).size
+      0.until(numCols).map { colIdx =>
+        rowStore.map(row => row(colIdx))
+      }.toArray
+    }
+  }
+
+  private[impl] def finalizeTree(
+      rootNode: Node,
+      algo: OldAlgo.Algo,
+      numClasses: Int,
+      numFeatures: Int,
+      parentUID: Option[String]): DecisionTreeModel = {
+    parentUID match {
+      case Some(uid) =>
+        if (algo == OldAlgo.Classification) {
+          new DecisionTreeClassificationModel(uid, rootNode, numFeatures = numFeatures,
+            numClasses = numClasses)
+        } else {
+          new DecisionTreeRegressionModel(uid, rootNode, numFeatures = numFeatures)
+        }
+      case None =>
+        if (algo == OldAlgo.Classification) {
+          new DecisionTreeClassificationModel(rootNode, numFeatures = numFeatures,
+            numClasses = numClasses)
+        } else {
+          new DecisionTreeRegressionModel(rootNode, numFeatures = numFeatures)
+        }
+    }
+  }
+
+  private[impl] def getPredict(impurityCalculator: ImpurityCalculator): Predict = {
+    val pred = impurityCalculator.predict
+    new Predict(predict = pred, prob = impurityCalculator.prob(pred))
+  }
+
+  /**
+   * On driver: Grow tree based on chosen splits, and compute new set of active nodes.
+   *
+   * @param oldPeriphery  Old periphery of active nodes.
+   * @param bestSplitsAndGains  Best (split, gain) pairs, which can be zipped with the old
+   *                            periphery.  These stats will be used to replace the stats in
+   *                            any nodes which are split.
+   * @param minInfoGain  Threshold for min info gain required to split a node.
+   * @return  New active node periphery.
+   *          If a node is split, then this method will update its fields.
+   */
+  private[impl] def computeActiveNodePeriphery(
```

Comment:
Can this be merged into the other logic?  Here, driver/worker logic is not separate, so they should be merged.